test/ServerSpec.hs

```diff
@@ -0,0 +1,112 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase                 #-}
+{-# LANGUAGE NoMonoLocalBinds           #-}
+{-# LANGUAGE NoMonomorphismRestriction  #-}
+{-# LANGUAGE DeriveAnyClass             #-}
+{-# LANGUAGE DerivingStrategies        #-}
+
+module ServerSpec where
+
+import Test.Hspec hiding (Arg)
+import Server
+import Weft.Types
+import Weft.Internal.Types
+import Weft.Generics.Resolve
+import Weft.Generics.Hydrate
+import Data.Aeson
+import GHC.Generics
+import Test.QuickCheck hiding (Args)
+import Data.ByteString.Char8
+
+newtype Id = Id Int 
+  deriving          (ToJSON, Generic)
+  deriving stock    (Show)
+  deriving newtype  (Eq, Ord, Arbitrary)
+  deriving Read via (Int)
+
+newtype Name = Name String deriving (Generic, Show, Eq, Ord, Arbitrary, ToJSON)
+
+data GqlQuery ts = GqlQuery
+    { getUser :: Magic ts (Arg "id" Id -> User ts)
+    , getAllUsers :: Magic ts [User ts]
+    } deriving (Generic)
+
+deriving instance AllHave Eq (GqlQuery ts)       => Eq (GqlQuery ts)
+
+data User ts = User
+  { userId         :: Magic ts (Arg "arg" (Maybe Int) -> Id)
+  , userName       :: Magic ts Name
+  , userBestFriend :: Magic ts (Arg "arg" (Maybe Int) -> User ts)
+  , userFriends    :: Magic ts [User ts]
+  } deriving (Generic)
+
+deriving instance AllHave Show (User ts)     => Show (User ts)
+deriving instance AllHave Eq (User ts)       => Eq (User ts)
+deriving instance AllHave ToJSON (User ts)   => ToJSON (User ts)
+
+getAllUsersTestString :: ByteString
+getAllUsersTestString = "query { \n getAllUsers { \n userId \n userName \n userFriends { \n userId \n userName \n } \n } \n } \n "
+
+getAllUsersTestQuery :: Either String (Gql GqlQuery () () 'Query)
+getAllUsersTestQuery = Right (Gql { query = Just (ANil
+                                             , GqlQuery { getUser = Nothing
```

Comment:
Nitpick: possible to get this down so it fits in 80 columns? my mind melts as soon as i need wordwrapping